### PR TITLE
matrix-authentication-service: init at 0.7.0

### PR DESCRIPTION
**Pull request by teutat3s**
## Description of changes
MAS (Matrix Authentication Service) is an OAuth 2.0 and OpenID Provider server for Matrix.

It has been created to support the migration of Matrix to an OpenID Connect (OIDC) based authentication layer as per [MSC3861](https://github.com/matrix-org/matrix-doc/pull/3861).
One can learn more about Matrix and OIDC at [areweoidcyet.com](https://areweoidcyet.com/).

https://github.com/matrix-org/matrix-authentication-service/releases/tag/v0.7.0

Vendoring the `Cargo.lock` is necessary because of a dependency they use: `opa-wasm-0.1.0` (pinned to a git repo) .
 
**Future work:**
For this to be really useful, a NixOS module will be needed.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/ma/matrix-authentication-service/package.nix [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  policies = stdenv.mkDerivation {
+    inherit src version;
+    pname = "matrix-authentication-service-policies";
+    sourceRoot = "source/policies";
+    nativeBuiltInputs = [ gnumake open-policy-agent ];
+    preBuild = ''
+      # Patch opa binary path
+      substituteInPlace Makefile \
+        --replace "opa" "${open-policy-agent}/bin/opa";
+    '';
```
Comment (NickCao) [inline pkgs/by-name/ma/matrix-authentication-service/package.nix:41]: This shall not be needed.

File: pkgs/by-name/ma/matrix-authentication-service/package.nix [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+      mkdir -p $out/share
+      mv dist $out/share/matrix-authentication-service-frontend
+      runHook postInstall
+    '';
+  };
+  policies = stdenv.mkDerivation {
+    inherit src version;
+    pname = "matrix-authentication-service-policies";
+    sourceRoot = "source/policies";
+    nativeBuiltInputs = [ gnumake open-policy-agent ];
```
Comment (NickCao) [inline pkgs/by-name/ma/matrix-authentication-service/package.nix:36]:
```suggestion
    nativeBuiltInputs = [ open-policy-agent ];
```

Isn't gnumake already part of stdenv?

File: pkgs/by-name/ma/matrix-authentication-service/package.nix [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      runHook preInstall
+      mkdir -p $out/share
+      mv dist $out/share/matrix-authentication-service-frontend
+      runHook postInstall
+    '';
+  };
+  policies = stdenv.mkDerivation {
+    inherit src version;
+    pname = "matrix-authentication-service-policies";
+    sourceRoot = "source/policies";
```
Comment (NickCao) [inline pkgs/by-name/ma/matrix-authentication-service/package.nix:35]:
```suggestion
    makeFlags = [ "-C" "policies" ];
```

File: pkgs/by-name/ma/matrix-authentication-service/package.nix [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+  cargoLock = {
+    lockFile = ./Cargo.lock;
+    outputHashes = {
+      "opa-wasm-0.1.0" = "sha256-GuOlUNGegdDieTvthk9SyfQSTeem7ArJTdiD1t7Ojd4=";
+    };
+  };
+
+  # Adopted from https://github.com/matrix-org/matrix-authentication-service/blob/main/Dockerfile
+  postInstall = ''
```
Comment (NickCao) [inline pkgs/by-name/ma/matrix-authentication-service/package.nix:66]: The paths in rust codes might also need to be patched.

**Review by NickCao — COMMENTED** [review]
Actually I've packaged this before, see https://github.com/NickCao/flakes/blob/master/pkgs/matrix-authentication-service/default.nix for some ideas to steal from.

**Comment by teutat3s** [comment]
> Actually I've packaged this before, see https://github.com/NickCao/flakes/blob/master/pkgs/matrix-authentication-service/default.nix for some ideas to steal from.

Thanks, I liked your version a lot and stole a lot (if not all :smile:).

**Review by NickCao — COMMENTED** [review]
I think we also need a nixos test for this, it's a complicated piece of software. Otherwise LGTM.

**Comment by teutat3s** [comment]
Do you mean a simple version test? In case you think of something more complex, can you point me to an example?

**Comment by NickCao** [comment]
> Do you mean a simple version test? In case you think of something more complex, can you point me to an example?

Something like `nixos/tests/matrix/synapse.nix`, but with MAS instead of builtin authentication.

**Comment by teutat3s** [comment]
Was missing `translations` path.

I started something regarding module / test in this branch: https://github.com/teutat3s/nixpkgs/commit/287fbae6e5bbc5baed1fef45c9497327f067c1d3

**Comment by adamcstephens** [comment]
@teutat3s are you planning to add the module to this PR or should we move forward with the package alone?

**Comment by teutat3s** [comment]
@adamcstephens I'm still working on the module and might be able to look into it some more in the coming week.
For me it would be fine to merge this now and add the module in a separate PR.

**Review by adamcstephens — APPROVED** [review]
Result of `nixpkgs-review pr 275057` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>matrix-authentication-service</li>
  </ul>
</details>

-----

Result of `nixpkgs-review pr 275057` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>matrix-authentication-service</li>
  </ul>
</details>

-----

Result of `nixpkgs-review pr 275057` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>matrix-authentication-service</li>
  </ul>
</details>

-----

Result of `nixpkgs-review pr 275057` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>matrix-authentication-service</li>
  </ul>
</details>

**Pull request merged**